gradle-plugin-testing-junit/src/main/java/com/palantir/gradle/testing/files/gradle/Plugins.java

```diff
@@ -0,0 +1,116 @@
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.testing.files.gradle;
+
+import com.google.errorprone.annotations.RestrictedApi;
+import com.palantir.gradle.testing.RestrictedCreation;
+import com.palantir.gradle.testing.files.ProjectFile.FileEditor;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Structured API for managing the {@code plugins {}} block in Gradle files.
+ * <p>
+ * Example: {@code rootProject.buildGradle().plugins().add("java").add("application")}
+ *
+ * @see GradleFile#plugins()
+ */
+public record Plugins(Function<FileEditor, ? extends GradleFile> editorFunction) {
+    private static final Pattern PLUGINS_BLOCK_PATTERN = Pattern.compile("plugins\\s*\\{([^}]*)}\\n?", Pattern.DOTALL);
+    private static final Pattern BUILDSCRIPT_BLOCK_PATTERN =
+            Pattern.compile("buildscript\\s*\\{(?:[^{}]|\\{[^{}]*})*}", Pattern.DOTALL);
+
+    @RestrictedApi(explanation = RestrictedCreation.EXPLANATION, allowedOnPath = RestrictedCreation.ALLOWED_ON_PATH)
+    public Plugins {}
+
+    /**
+     * Adds a plugin without a version. Example: {@code add("java")}
+     */
+    public Plugins add(String pluginId) {
+        return addPlugin(new Plugin(pluginId, Optional.empty(), Optional.empty()));
+    }
+
+    /**
+     * Adds a plugin with apply false. Example: {@code addWithoutApply("com.example.plugin")}
+     */
+    public Plugins addWithoutApply(String pluginId) {
+        return addPlugin(new Plugin(pluginId, Optional.empty(), Optional.of(false)));
+    }
+
+    private Plugins addPlugin(Plugin plugin) {
+        GradleFile unused = editorFunction.apply(text -> addPlugin(text, plugin));
+        return this;
+    }
+
+    private String addPlugin(String originalText, Plugin plugin) {
+        List<Plugin> plugins =
+                Stream.concat(parse(originalText).stream(), Stream.of(plugin)).toList();
+
+        return insertPluginsBlock(PLUGINS_BLOCK_PATTERN.matcher(originalText).replaceFirst(""), plugins);
+    }
+
+    private String insertPluginsBlock(String text, List<Plugin> plugins) {
+        String pluginsContent =
+                plugins.stream().map(plugin -> "    " + plugin.toDeclaration()).collect(Collectors.joining("\n"));
+
+        String pluginsBlock = "plugins {\n" + pluginsContent + "\n}\n";
+
+        Matcher buildscriptMatcher = BUILDSCRIPT_BLOCK_PATTERN.matcher(text);
+
+        if (buildscriptMatcher.find()) {
+            int insertPosition = buildscriptMatcher.end();
+            String suffix = text.substring(insertPosition).replaceFirst("^\n", "");
+            return text.substring(0, insertPosition) + "\n" + pluginsBlock + suffix;
+        }
+
+        return pluginsBlock + text;
+    }
+
+    static void assertNoManipulation(String oldText, String newText) {
```

Comment:
This needs a much better name. Manipulation of what? Perhaps `assertNoIncorrectPluginApplications`?